quinn-proto/src/congestion/cubic.rs

```diff
@@ -0,0 +1,264 @@
+    ssthresh: u64,
+    /// The time when QUIC first detects a loss, causing it to enter recovery. When a packet sent
+    /// after this time is acknowledged, QUIC exits recovery.
+    recovery_start_time: Option<Instant>,
+    cubic_state: State,
+}
+
+impl Cubic {
+    /// Construct a state using the given `config` and current time `now`
+    pub fn new(config: Arc<CubicConfig>, _now: Instant) -> Self {
+        Self {
+            window: config.initial_window,
+            ssthresh: u64::MAX,
+            recovery_start_time: None,
+            config,
+            cubic_state: Default::default(),
+        }
+    }
+}
+
+impl Controller for Cubic {
+    fn on_ack(
+        &mut self,
+        now: Instant,
+        sent: Instant,
+        bytes: u64,
+        app_limited: bool,
+        rtt: Duration,
+    ) {
+        if app_limited
+            || self
+                .recovery_start_time
+                .map(|recovery_start_time| sent <= recovery_start_time)
+                .unwrap_or(false)
+        {
+            return;
+        }
+
+        if self.window < self.ssthresh {
+            // Slow start
+            self.window += bytes;
+        } else {
+            // Congestion avoidance.
+            let ca_start_time;
+
+            match self.recovery_start_time {
+                Some(t) => ca_start_time = t,
+                None => {
+                    // When we come here without congestion_event() triggered,
+                    // initialize congestion_recovery_start_time, w_max and k.
+                    ca_start_time = now;
+                    self.recovery_start_time = Some(now);
+
+                    self.cubic_state.w_max = self.window as f64;
+                    self.cubic_state.k = 0.0;
+                }
+            }
+
+            let t = now - ca_start_time;
+
+            // w_cubic(t + rtt)
+            let w_cubic = self
+                .cubic_state
+                .w_cubic(t + rtt, self.config.max_datagram_size);
+
+            // w_est(t)
+            let w_est = self
+                .cubic_state
+                .w_est(t, rtt, self.config.max_datagram_size);
+
+            let mut cubic_cwnd = self.window;
+
+            if w_cubic < w_est {
+                // TCP friendly region.
+                cubic_cwnd = cmp::max(cubic_cwnd, w_est as u64);
+            } else if cubic_cwnd < w_cubic as u64 {
+                // Concave region or convex region use same increment.
+                let cubic_inc = (w_cubic - cubic_cwnd as f64) / cubic_cwnd as f64
+                    * self.config.max_datagram_size as f64;
+
+                cubic_cwnd += cubic_inc as u64;
+            }
+
+            // Update the increment and increase cwnd by MSS.
+            self.cubic_state.cwnd_inc += cubic_cwnd - self.window;
+
+            // cwnd_inc can be more than 1 MSS in the late stage of max probing.
+            // however QUIC recovery draft 7.4 (Congestion Avoidance) limits
```

Comment:
```suggestion
            // however RFC9002 §7.3.3 (Congestion Avoidance) limits
```

I'm not actually sure if the wording in that section should be considered applicable to non-NewReno congestion controllers. Does quiche clamp the window growth similarly?

Reply:
yes the only major difference with the quiche impl is the slow start. I kept the same slow start as quinn's new reno instead of implementing hystart.